lib/Sema/CSApply.cpp

```diff
@@ -6588,6 +6588,10 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
       if (toType->hasUnresolvedType())
         break;
 
+      if (cs.toImmutablePossible(fromType->lookThroughAllOptionalTypes(),
```

Comment:
Here in CSApply, we need to generate a correct type-checked AST, e.g., by calling the appropriate initializer on the target type. The handling of `ConversionRestrictionKind::CGFloatToDouble` is a good guide here.

Reply:
Also, we shouldn't be looking through optional types here. If there are optional mismatches, they'll have to be handled so the resulting type-checked AST is correct.